repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/PreProcessorUtils.java

```diff
@@ -198,4 +212,208 @@ public static void verifyDuplicateAssetByName(String typeName, String assetName,
+        mustArray.add(mapOf("term", mapOf(LEXICOGRAPHICAL_SORT_ORDER, lexoRank)));
+        if(StringUtils.isNotEmpty(glossaryQualifiedName)) {
+            mustArray.add(mapOf("terms", mapOf("__typeName.keyword", Arrays.asList("AtlasGlossaryTerm", "AtlasGlossaryCategory"))));
+            mustArray.add(mapOf("term", mapOf("__glossary", glossaryQualifiedName)));
+        } else{
+            mustArray.add(mapOf("terms", mapOf("__typeName.keyword", Arrays.asList("AtlasGlossary"))));
+        }
+
+        if(StringUtils.isEmpty(parentQualifiedName)) {
+            mustNotArray.add(mapOf("exists", mapOf("field", "__categories")));
+            mustNotArray.add(mapOf("exists", mapOf("field", "__parentCategory")));
+            boolFilter.put("must_not", mustNotArray);
+        }
+        else {
+            List<Map<String, Object>> shouldParentArray = new ArrayList<>();
+            shouldParentArray.add(mapOf("term", mapOf("__categories", parentQualifiedName)));
+            shouldParentArray.add(mapOf("term", mapOf("__parentCategory", parentQualifiedName)));
+            mustArray.add(mapOf("bool",mapOf("should", shouldParentArray)));
+        }
+
+        boolFilter.put("must", mustArray);
+
+        Map<String, Object> nestedBoolQuery = mapOf("bool", boolFilter);
+
+        Map<String, Object> topBoolFilter = mapOf("filter", nestedBoolQuery);
+
+        boolQuery.put("bool", topBoolFilter);
+
+        return boolQuery;
+    }
+    public static Map<String, Object> generateDSLQueryForLastCategory(String glossaryQualifiedName, String parentQualifiedName, boolean isTerm) {
+
+        Map<String, Object> sortKeyOrder = mapOf(LEXICOGRAPHICAL_SORT_ORDER, mapOf("order", "desc"));
+        Map<String, Object> scoreSortOrder = mapOf("_score", mapOf("order", "desc"));
+        Map<String, Object> displayNameSortOrder = mapOf("displayName.keyword", mapOf("order", "desc"));
+
+        Object[] sortArray = {sortKeyOrder, scoreSortOrder, displayNameSortOrder};
+
+        Map<String, Object> functionScore = mapOf("query", buildBoolQuery(glossaryQualifiedName, parentQualifiedName, isTerm));
+
+        Map<String, Object> dsl = new HashMap<>();
+        dsl.put("from", 0);
+        dsl.put("size", 1);
+        dsl.put("sort", sortArray);
+        dsl.put("query", mapOf("function_score", functionScore));
+
+        return dsl;
+    }
+
+    private static Map<String, Object> buildBoolQuery(String glossaryQualifiedName, String parentQualifiedName, boolean isTerm) {
+        Map<String, Object> boolQuery = new HashMap<>();
+        int mustArrayLength = 0;
+        if(StringUtils.isEmpty(parentQualifiedName) && StringUtils.isEmpty(glossaryQualifiedName)){
+            mustArrayLength = 2;
+        } else if(StringUtils.isEmpty(parentQualifiedName) && StringUtils.isNotEmpty(glossaryQualifiedName)){
+            mustArrayLength = 3;
+        } else {
+            mustArrayLength = 4;
+        }
+        Map<String, Object>[] mustArray = new Map[mustArrayLength];
```

Comment:
Avoid all this complexity by 
List<Map<String, Object>> mustList = new ArrayList<>();